DMIC: Add Kconfig options to reduce driver tables size

singalsu opened the pull request:
This patch adds two configuration profiles DMIC_FIR_LOW_MEMORY
and DMIC_FIR_FULL. The first is recommended for systems with
need for only 48 kHz and 16 kHz sample rates support with default
2.4 MHz microphone clock. It can work with some other configurations
also but if driver rejects are seen for desired parameters the
full configuration should used. The full configuration adds on top
of the first more supported decimation factors to support several
rates between 8 - 96 kHz.

Signed-off-by: Seppo Ingalsuo <seppo.ingalsuo@linux.intel.com>

singalsu commented:
The exposed kconfig may not be the cleanest possible. If you have recommendations please let me know! I first tried to get radio buttons type of selection between the two but ended up the make low RAM the base and full additional. So for full setup as we had previously both need to be selected. When de-selecting both there's "expert mode" to manually select any FIR decimation factors.

singalsu commented:
Also you can see that we have e.g. for decimation factor 2 two coefficient sets. The bandwidth of these sets differ. The lower bandwidth FIR is shorter that is needed in some cases to avoid FIR overrun. Also those could be exposed to kconfig to shrink further but it may be tricky to understand the purpose. For reaching absolute minimum it would be still beneficial. Please let me know what you think.

jajanusz commented on src/drivers/Kconfig:
```diff
@@ -7,4 +7,78 @@ config DMIC
+		preserve support for 48 kHz and 16 kHz sample rates
+		at typical 2.4 MHz microphone clock rate.
+
+config DMIC_FIR_FULL
```
I'd name that one something like ADDITIONAL/EXTRA, when I see CONFIG_*_FULL I expect it to have everything when it may not be a case here

lgirdwood replied:
@singalsu and also include CAVS prefix if Intel only code

jajanusz commented:
@singalsu 
Looks good, but we may tweak Kconfig a bit.
Please consider something like this:
https://gist.github.com/jajanusz/f08a0f03b0f47668fd39216799a6b9b2

zrombel commented:
Hi, CI Dmic test are failing on this PR for all platforms. For APL there's a IPC timeout and for CNL and ICL 8DEA0000 Notification when sending DaiConfig to FW. CI test are depending on <platform>_defconfig, does those files need any changes?

jajanusz commented:
@zrombel @singalsu 
It's weird, because all added configs are enabled by default, so probably some #if's are wrong/missing

singalsu commented:
@jajanusz Superb, thanks for the example! I will update this today. Seems like the saving is 4 kB so this looks worth the effort.

singalsu commented:
@jajanusz The pre-commit scripts gives me this error. Do you have idea what should be added to pass cleanly?

```
WARNING: please write a paragraph that describes the config symbol fully
#42: FILE: src/drivers/Kconfig:31:
+config DMIC_FIR_LOW_MEMORY

total: 0 errors, 1 warnings, 105 lines checked

```

jajanusz commented:
> @jajanusz The pre-commit scripts gives me this error. Do you have idea what should be added to pass cleanly?
> 
> ```
> WARNING: please write a paragraph that describes the config symbol fully
> #42: FILE: src/drivers/Kconfig:31:
> +config DMIC_FIR_LOW_MEMORY
> 
> total: 0 errors, 1 warnings, 105 lines checked
> ```

By 'paharagraph' checkpatch means 4 lines of text ;) in `help` for this symbol. It's guideline from kernel, it's more of a tip than, strict requirements I guess, for some configs it would be hard to make 4 lines long description.

singalsu commented:
@jajanusz OK, I'll update the PR. Thanks for help again!

singalsu commented:
Here's a new update. I noticed I had by mistake two commits in previous PR version, fixed that. I added some dots to get past the four lines check in pre-commit script and replaced spaces with tabs.

lgirdwood commented on src/drivers/Kconfig:
```diff
@@ -5,6 +5,112 @@ config DMIC
+		.
+		.
+
+if DMIC
```
This should probably be prefixed for INTEL_DMIC as there may be non Intel DMICs

lgirdwood replied:
or CAVS_DMIC

lgirdwood commented on src/drivers/Kconfig:
```diff
@@ -5,6 +5,112 @@ config DMIC
+menu "Decimation factors"
+	visible if DMIC_FIR_CUSTOM
+
+config DMIC_FIR_DECIMATE_BY_2
```
Ditto here for CAVS prefix

singalsu replied:
Yep, this is Intel and CAVS specific. I'll update the PR.

singalsu commented:
@jajanusz Sorry, I missed your comment about Linux kernel kconfig style (Documentation/kbuild/kconfig-language.txt), I'll do the update to indents as well.

singalsu commented:
I have some memory issue. DSP boots OK with low memory configuration, full and custom with most decimation factors enabled does DSP panic. I'm doing some tests with memory.h or try reduce stack usage if it could help. This patch should not grow DSP so it's strange.

singalsu commented:
@jajanusz @lgirdwood Not a memory issue, there was a potentially endless loop in previous version (just in PR, not in git master) that I've now fixed. I've now addressed your comments. I had to edit the config name change into many platform files. Can you please re-check this if the CI passes.

jajanusz commented on src/drivers/Kconfig:
```diff
@@ -1,13 +1,118 @@
+	help
+	  This option is used to select each supported decimation
+	  factor.
+	  .
```
No need to add these dots just to satisfy checkpatch, it's just guide, shorter description may be ok

singalsu replied:
OK, I'll make a new version. The warnings from them are shown only in post commit check so committing this was possible.

jajanusz commented:
@singalsu 
You can just `git commit --no-verify -s ... ` (please note no-verify has to be 1st arg), then it will commit successfully, BUT please use it only if you are sure that checkpatch overcomplaining

jajanusz commented:
> @singalsu can you check CI

Klocwork server is down, again ...

singalsu commented:
I'll check the CI issue with @zrombel . The defconfigs are now changed, I wonder if that impacts.

jajanusz commented:
> I'll check the CI issue with @zrombel . The defconfigs are now changed, I wonder if that impacts.

Tests are ok, just code scan (Klocwork) server was down, but should be up now

singalsu commented:
@lgirdwood Now all the 7 checks have been passed!
